### PR TITLE
[SYCL] Refactor node class

### PR DESCRIPTION
**Pull request by Bensuo**
- Node class now wraps a commad group object
- Some changes to handler::finalize to support this
- Remove refactored handler CG creation
- Minor changes to CG classes to support copying
- Simplify enqueueImpCommandBufferKernel parameters

### CODE REVIEW

File: sycl/source/detail/graph_impl.hpp [hunk 1]
```diff
@@ -37,35 +37,10 @@ class node_impl {
...
-
-  /// Store arg descriptors for the kernel arguments.
-  std::vector<sycl::detail::ArgDesc> MArgs;
-  /// We need to store local copies of the values pointed to by MArgs since they
-  /// may go out of scope before execution.
-  std::vector<std::vector<char>> MArgStorage;
-
-  /// Stores auxiliary resources used by internal operations.
-  std::vector<std::shared_ptr<const void>> MAuxiliaryResources;
+  // Command group object which stores all args etc needed to enqueue the node
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.hpp]:
```suggestion
  /// Command group object which stores all args etc needed to enqueue the node.
```

File: sycl/source/detail/graph_impl.hpp [hunk 2]
```diff
@@ -170,6 +113,69 @@ class node_impl {
   /// Query if this is an empty node.
   /// @return True if this is an empty node, false otherwise.
   bool is_empty() const { return MIsEmpty; }
+
+  std::unique_ptr<sycl::detail::CG> getCGCopy() const {
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.hpp:120]: Should have Doxygen comment

File: sycl/source/detail/graph_impl.hpp [hunk 3]
```diff
@@ -204,17 +210,9 @@ class graph_impl {
   /// @param DepEvents Dependent events of the created node.
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.hpp]: These doxygen params need updated

File: sycl/source/detail/graph_impl.hpp [hunk 4]
```diff
@@ -170,6 +113,69 @@ class node_impl {
...
+    case sycl::detail::CG::None:
+      assert(false &&
+             "Error: Empty nodes should not be enqueue to a command buffer");
+      break;
+    }
+    return nullptr;
+  }
+
+private:
+  template <typename CGT> std::unique_ptr<CGT> createCGCopy() const {
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.hpp:185]: doxygen comment

File: sycl/source/detail/graph_impl.hpp [hunk 5]
```diff
@@ -153,7 +93,10 @@ class node_impl {
   /// @param Arg Argument to lookup.
   /// @return True if \p Arg is used in node, false otherwise.
   bool has_arg(const sycl::detail::ArgDesc &Arg) {
-    for (auto &NodeArg : MArgs) {
+    // TODO: Handle types other than exec kernel
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.hpp:96]: Should we assert here on the type for now

File: sycl/source/handler.cpp [hunk 6]
```diff
@@ -1057,6 +1036,8 @@ void handler::ext_oneapi_graph(
     auto SubgraphEvent = std::make_shared<event_impl>();
     ParentGraph->add_event_for_node(SubgraphEvent, MSubgraphNode);
   } else {
+    MIsFinalized = true;
+    // TODO: This could probably be moved to handler::finalize()
```
Comment (EwanC) [inline sycl/source/handler.cpp]: How much work would it be to do this as part of this PR?

File: sycl/source/detail/graph_impl.hpp (hunk 5 above)
Comment (Bensuo) [reply sycl/source/detail/graph_impl.hpp:96]: Done!

File: sycl/source/detail/graph_impl.hpp (hunk 2 above)
Comment (Bensuo) [reply sycl/source/detail/graph_impl.hpp:120]: Done!

File: sycl/source/detail/graph_impl.hpp (hunk 4 above)
Comment (Bensuo) [reply sycl/source/detail/graph_impl.hpp:185]: Done!

File: sycl/source/detail/graph_impl.hpp (hunk 3 above)
Comment (Bensuo) [reply sycl/source/detail/graph_impl.hpp]: Done!

File: sycl/source/detail/graph_impl.hpp (hunk 1 above)
Comment (Bensuo) [reply sycl/source/detail/graph_impl.hpp]: Done!

File: sycl/source/handler.cpp (hunk 6 above)
Comment (Bensuo) [reply sycl/source/handler.cpp]: Turned out to be mostly trivial (just needs a new handler member) so I've done it now!

**Review by EwanC — APPROVED** [review]
LGTM - Less verbose implementation of node and don't need the busy new entry-point to the core handler class.

**Review by julianmi — APPROVED** [review]
Great improvements!

File: sycl/source/detail/graph_impl.cpp [hunk 7]
```diff
@@ -412,40 +389,36 @@ sycl::event exec_graph_impl::enqueue(
...
-          {} /* HostPipeName */, nullptr /* HostPipePtr */,
-          false /* HostPipeBlocking */, 0 /* HostPipeTypeSize */,
-          false /* HostPipeRead */, {} /* Advice */, NodeImpl->MRequirements,
-          {} /* Events */, {} /* Events w/ Barrier */,
-          NodeImpl->MOSModuleHandle,
-          PI_EXT_KERNEL_EXEC_INFO_CACHE_DEFAULT
-          /* KernelCacheConfig */,
-          {} /* CodeLoc */);
+
+      auto EventImpl = sycl::detail::Scheduler::getInstance().addCG(
```
Comment (reble) [inline sycl/source/detail/graph_impl.cpp]:
```suggestion
      sycl::detail::EventImplPtr EventImpl = sycl::detail::Scheduler::getInstance().addCG(
```

File: sycl/source/handler.cpp [hunk 8]
```diff
@@ -448,22 +268,184 @@ event handler::finalize() {
...
+  }
+
+  // If the queue has an associated graph then we need to take the CG and pass
+  // it to the graph to create a node, rather than submit it to the scheduler.
+  if (auto GraphImpl = MQueue->getCommandGraph(); GraphImpl) {
+    auto EventImpl = std::make_shared<detail::event_impl>();
+
+    // Extract relevant data from the handler and pass to graph to create a
+    // new node representing this command group.
+    auto NodeImpl = GraphImpl->add(MCGType, std::move(CommandGroup));
```
Comment (reble) [inline sycl/source/handler.cpp]:
```suggestion
    std::shared_ptr<node_impl> NodeImpl = GraphImpl->add(MCGType, std::move(CommandGroup));
```

File: sycl/source/handler.cpp [hunk 9]
```diff
@@ -448,22 +268,184 @@ event handler::finalize() {
...
+        MImpl->HostPipeTypeSize, MImpl->HostPipeRead, std::move(MArgsStorage),
+        std::move(MAccStorage), std::move(MSharedPtrStorage),
+        std::move(MRequirements), std::move(MEvents), MCodeLoc));
+    break;
+  }
+  case detail::CG::ExecCommandBuffer:
+    // If we have a subgraph node we don't want to actually execute this command
+    // graph submission.
+    if (!MSubgraphNode) {
+      auto GraphCompletionEvent = MExecGraph->exec(MQueue);
```
Comment (reble) [inline sycl/source/handler.cpp]:
```suggestion
      sycl::event GraphCompletionEvent = MExecGraph->exec(MQueue);
```

**Review by reble — APPROVED** [review]
LGTM

File: sycl/source/detail/graph_impl.cpp (hunk 7 above)
Comment (Bensuo) [reply sycl/source/detail/graph_impl.cpp]: Fixed!

File: sycl/source/handler.cpp (hunk 9 above)
Comment (Bensuo) [reply sycl/source/handler.cpp]: Fixed!

File: sycl/source/handler.cpp (hunk 8 above)
Comment (Bensuo) [reply sycl/source/handler.cpp]: Fixed!

**Pull request merged**